### PR TITLE
BUG: concat not copying index and columns when copy=True

### LINKED ISSUE #29879: pd.concat with copy=True doesn't copy columns index
#### Code Sample

```python
import pandas as pd
print(pd.__version__)  # '0.25.3'

# simple frame with a name attribute for columns
df = pd.DataFrame({'a': [1.1], 'b': [2.2]})
df.columns.name = 'x'
print(df)
# x    a    b
# 0  1.1  2.2

# tile rows to a new frame, apparently using the copy of the first frame
df2 = pd.concat([df] * 2, copy=True)
print(df2)
# x    a    b
# 0  1.1  2.2
# 0  1.1  2.2

# clear the column name for the first frame
df.columns.name = None

# and observe the name for the second
assert df2.columns.name is None
print(df2)
#      a    b
# 0  1.1  2.2
# 0  1.1  2.2

# aha, it's the same object
assert df.columns is df2.columns
```

#### Problem description

It is expected that `copy=True` should make a copy of the `columns` index. However, the demonstration shows it's the same object. Modifications, such as the `name` property, are made to the source frame and the output frame created by `pd.concat`.

#### Expected Output

It is expected that `df2.columns` is a copy of the original `df.columns`. This would prevent (e.g.) changing the `name` property for both frames.

A workaround is to manually copy this property after `pd.concat`:
```python
df2 = pd.concat([df] * 2, copy=True)
df2.columns = df2.columns.copy()
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.5.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.3
numpy            : 1.17.3
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 42.0.1.post20191125
Cython           : 0.29.14
pytest           : 5.3.0
hypothesis       : None
sphinx           : 2.2.1
blosc            : None
feather          : None
xlsxwriter       : 1.2.6
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.9.0
pandas_datareader: None
bs4              : 4.8.1
bottleneck       : 1.3.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.1
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.11
tables           : 3.6.1
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.6
</details>

### PR DESCRIPTION
**Pull request by fujiaxiang**
- [x] closes #29879
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by simonjayhawkins — COMMENTED** [review]
@fujiaxiang Thanks for the PR.

File: pandas/tests/reshape/test_concat.py [hunk 1]
```diff
@@ -2750,3 +2750,16 @@ def test_concat_sparse():
     )
     result = pd.concat([a, a], axis=1)
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("axis", [0, 1])
```
[the remark this answers is not in the document]
Comment (fujiaxiang) [reply pandas/tests/reshape/test_concat.py]: removed

File: pandas/tests/reshape/test_concat.py [hunk 2]
```diff
@@ -2750,3 +2750,16 @@ def test_concat_sparse():
...
+def test_concat_copy_index(axis):
+    # GH 29879
+    df = pd.DataFrame([[1, 2], [3, 4]], columns=["a", "b"])
+    df_comb = pd.concat([df, df], axis=axis, copy=True)
+    ser = df["a"]
+    ser_comb = pd.concat([ser, ser], axis=axis, copy=True)
```
[the remark this answers is not in the document]
Comment (fujiaxiang) [reply pandas/tests/reshape/test_concat.py]: done

File: doc/source/whatsnew/v1.0.0.rst [hunk 3]
```diff
@@ -1133,6 +1133,7 @@ Reshaping
 - Bug in :func:`merge_asof` merging on a tz-aware ``left_index`` and ``right_on`` a tz-aware column (:issue:`29864`)
 - Improved error message and docstring in :func:`cut` and :func:`qcut` when `labels=True` (:issue:`13318`)
 - Bug in missing `fill_na` parameter to :meth:`DataFrame.unstack` with list of levels (:issue:`30740`)
+- Bug in :func:`concat` index and columns not copied when ``copy=True`` (:issue:`29879`)
```
[the remark this answers is not in the document]
Comment (fujiaxiang) [reply doc/source/whatsnew/v1.0.0.rst]: done

File: pandas/core/reshape/concat.py [hunk 4]
```diff
@@ -516,9 +516,12 @@ def _get_new_axes(self) -> List[Index]:
 
     def _get_comb_axis(self, i: int) -> Index:
         data_axis = self.objs[0]._get_block_manager_axis(i)
-        return get_objs_combined_axis(
+        comb_axis = get_objs_combined_axis(
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: the copy arg should be pushed down thru to get_objs_combined_axis (it goes deeper, keep pushing it down)

File: pandas/core/reshape/concat.py [hunk 5]
```diff
@@ -516,9 +516,12 @@ def _get_new_axes(self) -> List[Index]:
...
             self.objs, axis=data_axis, intersect=self.intersect, sort=self.sort
         )
+        # GH 29879
+        # No need for deep copy. Indexes are immutable so they can share underlying data
+        return comb_axis.copy(deep=False) if self.copy else comb_axis
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: we *almost* never actually deep copy things. Why do you think we actually need to make this change?

Comment (jreback) [reply pandas/core/reshape/concat.py]: indexes are immutable and sharing them is ok.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/concat.py (hunk 4 above)
Comment (fujiaxiang) [reply pandas/core/reshape/concat.py]: done

File: pandas/core/reshape/concat.py (hunk 5 above)
Comment (fujiaxiang) [reply pandas/core/reshape/concat.py]: As mentioned by @jorisvandenbossche in #29879, it is the underlying data of indexes are considered immutable, the index object itself has some attributes (e.g. name) that can be mutated. We probably don't want changing the name of index of the concatenated object affect the original object even when ``copy=True``.

Have updated my code comment to be more precise.

File: doc/source/whatsnew/v1.1.0.rst [hunk 6]
```diff
@@ -144,7 +144,7 @@ Reshaping
 - Bug in :meth:`DataFrame.pivot_table` when only MultiIndexed columns is set (:issue:`17038`)
 - Fix incorrect error message in :meth:`DataFrame.pivot` when ``columns`` is set to ``None``. (:issue:`30924`)
 - Bug in :func:`crosstab` when inputs are two Series and have tuple names, the output will keep dummy MultiIndex as columns. (:issue:`18321`)
-
+- Bug in :func:`concat` index and columns not copied when ``copy=True`` (:issue:`29879`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: Bug in ....concat, where the resulting indices where not copied when ``copy=True`` (the default).

File: pandas/core/indexes/api.py [hunk 7]
```diff
@@ -143,6 +150,13 @@ def _get_combined_index(
             index = index.sort_values()
         except TypeError:
             pass
+
+    # GH 29879
+    if copy:
+        # No need for deep copy. The underlying data of Indexes are
+        # considered immutable so they can be shared
+        index = index.copy(deep=False)
```
Comment (jreback) [inline pandas/core/indexes/api.py]: you don't need the comment or deep=False which is the default.

**Review by jreback — CHANGES_REQUESTED** [review]
ok, looks fine, ping when green.

File: pandas/core/indexes/api.py (hunk 7 above)
Comment (fujiaxiang) [reply pandas/core/indexes/api.py]: ok then. I wanted to be a bit more explicit, but sure if you think it's unnecessary.

Have cleaned it up.

File: doc/source/whatsnew/v1.1.0.rst (hunk 6 above)
Comment (fujiaxiang) [reply doc/source/whatsnew/v1.1.0.rst]: updated

**Comment by fujiaxiang** [comment]
ping

**Comment by simonjayhawkins** [comment]
Thanks @fujiaxiang

**Pull request merged**